Danki2/Assets/Scripts/Abilities/AbilityBuilder.cs

```diff
@@ -0,0 +1,45 @@
+﻿using Abilities;
+using Assets.Scripts.Abilities.InstantCast;
+using System;
+using System.Collections.Generic;
+
+namespace Assets.Scripts.Abilities
+{
+    public enum AbilityReference
+    {
+        Slash,
+        ShieldBash,
+    }
+
+    public class AbilityBuilder
+    {
+        private readonly Func<AbilityContext, Ability> _buildAction;
+
+        public AbilityBuilder(Func<AbilityContext, Ability> buildAction)
+        {
+            _buildAction = buildAction;
+        }
+
+        public Ability Build(AbilityContext context)
+        {
+            return _buildAction(context);
+        }
+
+        // This is where we tie the AbilityReference enum to abilities.
+        private static Dictionary<AbilityReference, AbilityBuilder> abilityBuilders = new Dictionary<AbilityReference, AbilityBuilder>
+        {
+            { AbilityReference.Slash, new AbilityBuilder(c => new Slash(c)) },
+            { AbilityReference.ShieldBash, new AbilityBuilder(c => new ShieldBash(c)) }
+        };
+
+        public static AbilityBuilder GetAbilityBuilder(AbilityReference reference)
```

Comment:
A problem I can see is not knowing whether you've got an instant cast or a channel builder. So when you call .Build() to get your ability, you don't know the type and you'll have to do some type casting, maybe that's OK though and we can just handle that further down stream? I'm not sure but there might be a nicer way to solve this problem with generics?

Reply:
Addressed this in latest commit.